aggregator/aggregator.go

```diff
@@ -227,18 +270,31 @@ func (agg *aggregator) AddMetricWithPoliciesList(
 	return err
 }
 
-func (agg *aggregator) Close() error {
+func (agg *aggregator) Close(closeType CloseType) error {
 	agg.Lock()
-	defer agg.Unlock()
-
 	if agg.state != aggregatorOpen {
-		return errAggregatorIsNotOpenOrClosed
+		agg.Unlock()
+		return errAggregatorExpectOpen
+	}
+	if closeType == GracefulClose {
+		agg.state = aggregatorClosing
+		agg.Unlock()
+		ctx, cancel := context.WithTimeout(context.Background(), agg.gracefulCloseTimeout)
+		defer cancel()
+		if err := agg.electionManager.Resign(ctx); err != nil {
```

Comment:
Yes the semantics of the `Resign` API on the election manager is that the replica may only resign successfully and become a follower if it can confirm the other replica has successfully taken over the ownership. If both the leader and the follower race to resign, only one of them will win and become the follower and the other replica will fail to resign and assume the leader role, until the follower is back up and ready to take over the leader role.

Additionally, we'll ensure no two replicas owning the same shardset will ever be deployed at the same time in the deployment tooling for the agg tier.